### ISSUE TITLE
perl jumps parens, overwriting original copy

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#13425]
Migrated from [rt.perl.org#120553](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=120553) (status was 'rejected')

Searchable as RT120553$

**Comment by p5pRT** [comment]
### From perl-diddler@tlinx.org

#### Created by perl-diddler@tlinx.org


Most curious&#8203;:


\#\!/usr/bin/perl \-w
use strict;

our $sep="/";

my \($x\, $y\) = \("this"\, "that"\);

$y =~ \(s\!&#8203;::\!$sep\!gee\);   $y \.= "\.suffix";
printf "y=%s\\n"\, $y;

$x = \($x =~ s\!&#8203;::\!$sep\!gee\) \. "\.suffix";
printf "x=%s\\n"\, $x;

\-\-\-\-\-\-\-\-\-cut\-\-\-\-\-\-
the case with 'x' and 'this' looks like an optimization of the 
one with 'y' and 'that'\.  Unfortunately perl is "excessively 
optimizing"\[understatement\] 'this' out of the result&#8203;:  I\.e\.
running the above produces&#8203;:
> x\.pl   
y=that\.suffix
x=\.suffix


What happened?

If you claim this is a feature\, \(\*cough\*\)\,
then what would be the correct way to  combine
them into 1 statement?

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=high

This perlbug was built using Perl 5.18.1 - Fri Sep 27 18:27:33 UTC 2013
It is being executed now by  Perl 5.16.2 - Sat Jan 26 21:52:41 UTC 2013.

Site configuration information for perl 5.16.2:

Configured by abuild at Sat Jan 26 21:52:41 UTC 2013.

Summary of my perl5 (revision 5 version 16 subversion 2) configuration:
   
  Platform:
    osname=linux, osvers=3.4.6-2.10-default, archname=x86_64-linux-thread-multi
    uname='linux build35 3.4.6-2.10-default #1 smp thu jul 26 09:36:26 utc 2012 (641c197) x86_64 x86_64 x86_64 gnulinux '
    config_args='-ds -e -Dprefix=/usr -Dvendorprefix=/usr -Dinstallusrbinperl -Dusethreads -Di_db -Di_dbm -Di_ndbm -Di_gdbm -Dd_dbm_open -Duseshrplib=true -Doptimize=-fmessage-length=0 -O2 -Wall -D_FORTIFY_SOURCE=2 -fstack-protector -funwind-tables -fasynchronous-unwind-tables -g -Wall -pipe -Accflags=-DPERL_USE_SAFE_PUTENV -Dotherlibdirs=/usr/lib/perl5/site_perl'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DPERL_USE_SAFE_PUTENV -fno-strict-aliasing -pipe -fstack-protector -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-fmessage-length=0 -O2 -Wall -D_FORTIFY_SOURCE=2 -fstack-protector -funwind-tables -fasynchronous-unwind-tables -g -Wall -pipe',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DPERL_USE_SAFE_PUTENV -fno-strict-aliasing -pipe -fstack-protector'
    ccversion='', gccversion='4.7.2 20130108 [gcc-4_7-branch revision 195012]', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib64 -fstack-protector'
    libpth=/lib64 /usr/lib64 /usr/local/lib64
    libs=-lm -ldl -lcrypt -lpthread
    perllibs=-lm -ldl -lcrypt -lpthread
    libc=/lib64/libc-2.17.so, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.17'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E -Wl,-rpath,/usr/lib/perl5/5.16.2/x86_64-linux-thread-multi/CORE'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib64 -fstack-protector'

Locally applied patches:
    


@INC for perl 5.16.2:
    /home/law/bin/lib
    /usr/lib/perl5/site_perl/5.16.2/x86_64-linux-thread-multi
    /usr/lib/perl5/site_perl/5.16.2
    /usr/lib/perl5/vendor_perl/5.16.2/x86_64-linux-thread-multi
    /usr/lib/perl5/vendor_perl/5.16.2
    /usr/lib/perl5/5.16.2/x86_64-linux-thread-multi
    /usr/lib/perl5/5.16.2
    /usr/lib/perl5/site_perl/5.16.2/x86_64-linux-thread-multi
    /usr/lib/perl5/site_perl/5.16.2
    /usr/lib/perl5/site_perl
    .


Environment for perl 5.16.2:
    HOME=/home/law
    LANG=en_US.utf8
    LANGUAGE (unset)
    LC_COLLATE=C
    LC_CTYPE=en_US.utf8
    LD_LIBRARY_PATH=/usr/lib64/mpi/gcc/openmpi/lib64
    LOGDIR (unset)
    PATH=/usr/lib/perl5/5.18/usr/bin/:/home/law/bin/lib:/sbin:/usr/local/sbin:/usr/lib64/mpi/gcc/openmpi/bin:/home/law/bin:/usr/local/bin:/usr/bin:/bin:/usr/bin/X11:/usr/X11R6/bin:/usr/games:/opt/kde3/bin:/usr/lib/mit/bin:/usr/lib/mit/sbin:.:/usr/lib/qt3/bin:/opt/dell/srvadmin/bin:/usr/sbin:/etc/local/func_lib:/home/law/lib
    PERL5OPT=-Mutf8 -CSA -I/home/law/bin/lib
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From perl-diddler@tlinx.org

Can't even close your own bugs\.\.\. that's atypical for a bug system\, but not for a heavy\-handed control system\.

Anyway\.\.\. nevermind\.  

You can close this bug\.

**Comment by p5pRT** [comment]
### From @emazep

On 2013\-11\-17 00&#8203;:53\, Linda Walsh \(via RT\) wrote&#8203;:
> \# New Ticket Created by  Linda Walsh 
> \# Please include the string&#8203;:  \[perl \#120553\]
> \# in the subject line of all future correspondence about this issue\. 
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=120553 >
> 
> 
> 
> This is a bug report for perl from perl\-diddler@&#8203;tlinx\.org\,
> generated with the help of perlbug 1\.39 running under perl 5\.16\.2\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> Most curious&#8203;:
> 
> 
> \#\!/usr/bin/perl \-w
> use strict;
> 
> our $sep="/";
> 
> my \($x\, $y\) = \("this"\, "that"\);
> 
> $y =~ \(s\!&#8203;::\!$sep\!gee\);   $y \.= "\.suffix";
> printf "y=%s\\n"\, $y;
> 
> $x = \($x =~ s\!&#8203;::\!$sep\!gee\) \. "\.suffix";
> printf "x=%s\\n"\, $x;
> 
> \-\-\-\-\-\-\-\-\-cut\-\-\-\-\-\-
> the case with 'x' and 'this' looks like an optimization of the 
> one with 'y' and 'that'\.  Unfortunately perl is "excessively 
> optimizing"\[understatement\] 'this' out of the result&#8203;:  I\.e\.
> running the above produces&#8203;:
>> x\.pl   
> y=that\.suffix
> x=\.suffix
> 
> 
> What happened?

I see nothing strange\.

Here&#8203;:

$x =~ s\!&#8203;::\!$sep\!gee

s/// does not find the pattern to replace\, so it returns false\, more
specifically the empty string\, which is then concatenated with
"\.suffix"\, and the result of the concatenation is assigned to $x\.

In other words\, the following line from your code&#8203;:

$x = \($x =~ s\!&#8203;::\!$sep\!gee\) \. "\.suffix";

simply reads&#8203;:

$x = '' \. "\.suffix";

\-Emanuele

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
@jkeenan - Status changed from 'open' to 'rejected'

**Comment by p5pRT** [comment]
### From zefram@fysh.org

Linda Walsh wrote&#8203;:
>What happened?

Your concatenation used the truth value returned by s///\.  You can add
the /r flag to make it return the updated subject string instead of the
truth value\.  However\, you have another problem&#8203;:

>$x = \($x =~ s\!&#8203;::\!$sep\!gee\) \. "\.suffix";

You're using /ee\, which requests two layers of evaluation of the
replacement\.  In this case\, one evaluation makes no difference compared
to the default interpolation behaviour\, but the second evaluation will
attempt to evaluate "/"\, if a substitution is ever made\.  That won't work\.
You don't want any /e flags here\.

So the correct statement would be&#8203;:

 $x = \($x =~ s\!&#8203;::\!$sep\!gr\) \. "\.suffix";

\-zefram

**Comment by p5pRT** [comment]
### From @Leont

On Sun\, Nov 17\, 2013 at 12&#8203;:53 AM\, Linda Walsh \<perlbug\-followup@&#8203;perl\.org>wrote&#8203;:

> Most curious&#8203;:
>
>
> \#\!/usr/bin/perl \-w
> use strict;
>
> our $sep="/";
>
> my \($x\, $y\) = \("this"\, "that"\);
>
> $y =~ \(s\!&#8203;::\!$sep\!gee\);   $y \.= "\.suffix";
> printf "y=%s\\n"\, $y;
>
> $x = \($x =~ s\!&#8203;::\!$sep\!gee\) \. "\.suffix";
> printf "x=%s\\n"\, $x;
>
> \-\-\-\-\-\-\-\-\-cut\-\-\-\-\-\-
> the case with 'x' and 'this' looks like an optimization of the
> one with 'y' and 'that'\.  Unfortunately perl is "excessively
> optimizing"\[understatement\] 'this' out of the result&#8203;:  I\.e\.
> running the above produces&#8203;:
> > x\.pl
> y=that\.suffix
> x=\.suffix
>
>
> What happened?
>

Linda\, can you please stop using this bugtracker as a general purpose Q&A
resource? There are plenty of other places where you can ask them\, you know
very well where to find perlmonks\.


> If you claim this is a feature\, \(\*cough\*\)\,
> then what would be the correct way to  combine
> them into 1 statement?
>

We claim it's your misunderstanding of the language\, which is ok except
when you blame us instead of yourself that this doesn't do what you expect
it to do\.

Leon

**Comment by p5pRT** [comment]
### From perl-diddler@tlinx.org

Leon\, if only you just had a few more clues\.\.\.I'm sure you wouldn't come off as nearly so slow\.  

Leon blathered&#8203;:
>  We claim it's your misunderstanding of the language\, which is ok except when you blame us instead of yourself that this doesn't do what you expect it to do\.


---
Hey\, I \*know\* I am using a buggy language that you won't fix in any released version  w/r/t mmap\, which causes perl to run 50% slower\.  And I know you are even clueless about the speed impact\.\.\.but running tests w/o imap take 300 seconds that running w/imap take 200\.\.\.

Furthermore\, I know I can't build & test a version of perl on my machine \(dating back to last year in notes on opensuse lists\) due to a bug in the \*gnu\-DB\* routines that causes the DB\-related tests to fail\(\*\*1\) \(which the gnu\-db folks were informed about 2 months ago\, but were too busy to fix before suse's 13\.1 w/the buggy perl became 'golden dung'&#8203;: ready for ship\.

Of course perl could patch their usage of the libraries as well\, but I didn't even bother to suggest that\, due to previous experiences getting you to fix known bugs\.\.\.

\(\*\*1\) \- http&#8203;://lists\.opensuse\.org/opensuse\-factory/2013\-09/msg00147\.html\.

But this ticket was already "closed" by me \(to the extent that I am permitted to do so\) less than 10 minutes after it was filed&#8203;:

Sat Nov 16 15&#8203;:53&#8203;:48 2013 Linda Walsh \<perl\-diddler@&#8203;tlinx\.org> \- Ticket created \.


---
Then\, less than 10 minutes later&#8203;:
Sat Nov 16 16&#8203;:02&#8203;:38 2013 Linda Walsh \<perl\-diddler@&#8203;tlinx\.org> \- Comments added \[Reply\] \[Comment\]
Download \(untitled\) / with headers
text/plain 161b
Can't even close your own bugs\.\.\. that's atypical for a bug system\, but not for a heavy\-handed control system\. Anyway\.\.\. nevermind\. You can close this bug\. 


---


---
So like I said\.  I \*\*know\*\* I am using a perl that is known to be buggy in multiple ways \-\- and that's not even including the multiple ways perl has mishandled UTF\-8 over the past decade rather than having people fix the bugs in their source code\.\.\. \(i\.e\.  forcing opening w/binary when they don't mean text\, something Windows users have had to do from the star\.  Then not only would perl have had consistent UTF\-8 treatment and no perl\-unicode bug but it would have been more likely to ensure binary mode where it was needed\.  Such treatment is necessary for routines to work on Windows ANYWAY\, so it would have cleaned up a mess of code\.  But rather than force cleaning up such code\, people spend years coding around the perl unicode bug for "real old legacy programs"\.

So please get\, that my attitude towards perl is mixed\, knowing that it is a great language\, with a flakey implementation\.\.\.that I try to work around at every known opportunity\, but having to deal with something with so many known but unfixed bugs does make me suspect perl a bit more than I might\, otherwise\.  This might be especially when true after wasting days dealing with double UTF\-8 encodings during test devel due to routines that were coded to get around the perl\-unicode bug\.\.\.\.

So I'll stop thinking perl might be at fault as soon as it fixes it's known bugs \(regardless of them being kept for bckwards compat or not\. \-\- not really an issue given that 5\.18 turns on warnings for previous non\-experimental features that cause fatal run\-time errors \-\- i\.e\. if previous compat was really an issue\, that wouldn't have been done\)\.

There is a history of incompatible treatment of compatibility \-\- keeping bugs that suit some\, while introducing new ones w/o the notice of the new process that was hilighted as a reason why the unicode bug wouldn't happen again \(despite it never being fixed \-\- though it moves closer to being so\.\.\.\)\.

Finally\, note that I would never have written this if you noticed that this bug was already closed by me less than 10 minutes after it being opened\, and you not taking the opportunity to respond such that you \*ask\* for this response\.

**Comment by p5pRT** [comment]
### From @rjbs

\* Linda Walsh via RT \<perlbug\-followup@&#8203;perl\.org> \[2013\-11\-17T15&#8203;:41&#8203;:09\]
> Leon\, if only you just had a few more clues\.\.\.I'm sure you wouldn't come off
> as nearly so slow\.  
> 
> Leon blathered&#8203;:
> \[\.\.\.\]

Linda\, you have worked very hard at becoming a complete persona non grata here\,
with smashing success\.

I have said "no" to every request to ban you from RT\, and there have been
plenty\.  I don't know why you can't see how your behavior is entirely
unacceptable\, but efforts have been made to politely point out that it is\.
Better than half your messages posted here result in private messages to me
imploring me to just ban you already\.  Let me note\, too that I did \*not\*
receive such from Leon\.

Learn how to hold your tongue\.  Even if you chose to read Leon's comment as
irritated\, you may note that he did not choose to insult you as you have chosen
to insult him\.

I'm getting really tired of watching the fallout in nearly every thread to
which you post\.

> Finally\, note that I would never have written this if you noticed that this
> bug was already closed by me less than 10 minutes after it being opened\, and
> you not taking the opportunity to respond such that you \*ask\* for this
> response\.

You added a comment to the bug\, rather than a reply\, which means the update was
not sent to the list\.  I agree that RT is imperfect and can be confusing even
to those of us who use it daily\, but Leon was acting in good faith by replying
to what he got on the mailing list\, and not jumping the gun by skipping further
mails\.  There were none\.

\-\- 
rjbs
